src/workerd/io/trace.c++

```diff
@@ -711,6 +783,549 @@ tracing::Exception::Exception(rpc::Trace::Exception::Reader reader)
+    : traceId(traceId),
+      invocationId(invocationId),
+      spanId(spanId) {}
+
+tracing::TailEvent::Context::Context(rpc::InvocationSpanContext::Reader reader)
+    : traceId(TraceId::fromCapnp(reader.getTraceId())),
+      invocationId(TraceId::fromCapnp(reader.getInvocationId())),
+      spanId(reader.getSpanId()) {}
+
+void tracing::TailEvent::Context::copyTo(rpc::InvocationSpanContext::Builder builder) {
+  traceId.toCapnp(builder.initTraceId());
+  invocationId.toCapnp(builder.initInvocationId());
+  builder.setSpanId(spanId);
+}
+
+tracing::TailEvent::Context tracing::TailEvent::Context::clone() {
+  return Context(traceId, invocationId, spanId);
+}
+
+tracing::TailEvent::TailEvent(kj::Rc<tracing::InvocationSpanContext>& context,
+    kj::Date timestamp,
+    kj::uint sequence,
+    Event&& event)
+    : context(getContextFromSpan(context)),
+      parentContext(getParentContextFromSpan(context)),
+      timestamp(timestamp),
+      sequence(sequence),
+      event(kj::mv(event)) {}
+
+tracing::TailEvent::TailEvent(Context context,
+    kj::Maybe<Context> parentContext,
+    kj::Date timestamp,
+    kj::uint sequence,
+    Event&& event)
+    : context(kj::mv(context)),
+      parentContext(kj::mv(parentContext)),
+      timestamp(timestamp),
+      sequence(sequence),
+      event(kj::mv(event)) {}
+
+namespace {
+tracing::TailEvent::Context readContextFromTailEvent(rpc::Trace::TailEvent::Reader& reader) {
+  return tracing::TailEvent::Context(reader.getContext());
+}
+
+kj::Maybe<tracing::TailEvent::Context> readParentContextFromTailEvent(
+    rpc::Trace::TailEvent::Reader& reader) {
+  if (!reader.hasParentContext()) return kj::none;
+  return tracing::TailEvent::Context(reader.getParentContext());
+}
+
+tracing::TailEvent::Event readEventFromTailEvent(rpc::Trace::TailEvent::Reader& reader) {
+  auto event = reader.getEvent();
+  switch (event.which()) {
+    case rpc::Trace::TailEvent::Event::ONSET:
+      return tracing::Onset(event.getOnset());
+    case rpc::Trace::TailEvent::Event::OUTCOME:
+      return tracing::Outcome(event.getOutcome());
+    case rpc::Trace::TailEvent::Event::SPAN_OPEN:
+      return tracing::SpanOpen(event.getSpanOpen());
+    case rpc::Trace::TailEvent::Event::SPAN_CLOSE:
+      return tracing::SpanClose(event.getSpanClose());
+    case rpc::Trace::TailEvent::Event::ATTRIBUTE:
+      return tracing::Mark(tracing::Attribute(event.getAttribute()));
+    case rpc::Trace::TailEvent::Event::RETURN:
+      return tracing::Mark(tracing::Return(event.getReturn()));
+    case rpc::Trace::TailEvent::Event::DIAGNOSTIC_CHANNEL_EVENT:
+      return tracing::Mark(tracing::DiagnosticChannelEvent(event.getDiagnosticChannelEvent()));
+    case rpc::Trace::TailEvent::Event::EXCEPTION:
+      return tracing::Mark(tracing::Exception(event.getException()));
+    case rpc::Trace::TailEvent::Event::LOG:
+      return tracing::Mark(tracing::Log(event.getLog()));
+  }
+  KJ_UNREACHABLE;
```

Comment:
Can you move this to the default case of the switch?

Reply:
Just Say No The `default:` Case.